README.md
# MVC JS Scaffolding Tool

<img src="logo.png">

MVC JS Scaffolding Tool is a VS Code Extention designed for a very specific architecture. Designed for internal use of a private company's web development process, this may not be the tool for you but I decided to make it public as a learning resource.

CHANGELOG.md
# Change Log

All notable changes to the "mvc-generator" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/Views/MVCSettings.ts
import * as vscode from 'vscode'
import saveSettings from '../UseCases/saveSettings'

class MVCSettings {
  public static currentPanel: MVCSettings | undefined
  public static readonly viewType = 'MVCSettings'
  private readonly _panel: vscode.WebviewPanel
  private readonly _extentionUri: vscode.Uri
  private _disposables: vscode.Disposable[] = []
  private context: vscode.ExtensionContext
  
  public static show (extentionUri: vscode.Uri, context: vscode.ExtensionContext) {
    const column = vscode.window.activeTextEditor
      ? vscode.window.activeTextEditor.viewColumn
      : undefined
    
      if (MVCSettings.currentPanel) {
        MVCSettings.currentPanel._panel.reveal(column)
      }

      const panel = vscode.window.createWebviewPanel(
        MVCSettings.viewType,
        'MVC Form',
        column || vscode.ViewColumn.One,
        {
          enableScripts: true,
          localResourceRoots: [vscode.Uri.joinPath(extentionUri, 'viewResources')]
        }
      )

      MVCSettings.currentPanel = new MVCSettings(panel, extentionUri, context)
  }

  public static revive (panel: vscode.WebviewPanel, extentionUri: vscode.Uri, context: vscode.ExtensionContext) {
    MVCSettings.currentPanel = new MVCSettings(panel, extentionUri, context)
  }

  private constructor (panel: vscode.WebviewPanel, extentionUri: vscode.Uri, context: vscode.ExtensionContext) {
    this._panel = panel
    this._extentionUri = extentionUri
    this.context = context

    this._update()

    this._panel.onDidDispose(() => this.dispose(), null, this._disposables)

    this._panel.onDidChangeViewState((e: any) => {
      if (this._panel.visible) this._update()
    },
    null,
    this._disposables
    )
  }

  public dispose () {
    MVCSettings.currentPanel = undefined
    this._panel.dispose()
    
    while (this._disposables.length) {
      const x = this._disposables.pop()
      if (x) x.dispose()
    }
  }

  private _update () {
    const webview = this._panel.webview
    
    this._panel.title = 'MVC Settings'
    this._panel.webview.html = this._getHtmlForWebview(webview)
    this._panel.webview.onDidReceiveMessage((e: any) => {
      saveSettings(e, this.context)
    },
    undefined,
    this._disposables
    )
  }

  private _getHtmlForWebview(webview: vscode.Webview) {
    const modelsDirectory = this.context.globalState.get('modelsDirectory')
    const modelTestsDirectory = this.context.globalState.get('modelTestsDirectory')
    return `
    <!DOCTYPE html>
    <html lang="en">

    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>MVC Settings</title>

      <style>
        button {
          color: white;
          background-color: rgb(75, 147, 255);
          border: solid rgb(49, 139, 255) 1px;
          border-radius: 4px;
          padding: 6px 10px;
          display: block;
          margin: 10px 0px;
        }

        label {
          display: block;
          margin-top: 6px;
        }

        input {
          display: block;
          background-color: rgba(255, 255, 255, 0.1);
          color: rgb(250, 250, 250);
          font-size: 16px;
          padding: 6px 10px;
          border: solid 1px rgb(196, 196, 196);
          margin-bottom: 6px;
        }
      </style>
    </head>

    <body>
      <label for="modelsDirectory">Models Directory</label>
      <input id="modelsDirectory" />

      <label for="modelTestsDirectory">Model Tests Directory</label>
      <input id="modelTestsDirectory" />


      <button onclick="submit()">Save</button>
    </body>

    <script>
      const vscode = acquireVsCodeApi()

      document.getElementById('modelsDirectory').value = '${modelsDirectory}'
      document.getElementById('modelTestsDirectory').value = '${modelTestsDirectory}'
      
      function submit() {
        vscode.postMessage(
          {
            modelsDirectory: document.getElementById('modelsDirectory').value,
            modelTestsDirectory: document.getElementById('modelTestsDirectory').value
          }
        )
      }
    </script>

    </html>
    `
  }
}

export default MVCSettings



src/UseCases/createModel.ts
import * as fs from 'fs'
import * as vscode from 'vscode'

interface ModelInput {
  name: string,
  propertyNames: string[]
}

function createModel (model: ModelInput, context: vscode.ExtensionContext) {
  const workspaceDirectory: string = vscode.workspace.rootPath || ''

  const modelsDirectory = `${workspaceDirectory}/${context.globalState.get('modelsDirectory')}`
  try {
    const doesModelsDirectoryExist = fs.existsSync(modelsDirectory)
    if (!doesModelsDirectoryExist) fs.mkdirSync(modelsDirectory)
  } catch {
    vscode.window.showErrorMessage('Ensure path to Models directory exists and is set in Settings')
    return
  }

  const modelToCreateDirectory = `${modelsDirectory}/${model.name}`
  const doesMoodelToCreateDirectoryExist = fs.existsSync(modelToCreateDirectory)
  if (!doesMoodelToCreateDirectoryExist) fs.mkdirSync(modelToCreateDirectory)

  const modelToCreateFilePath = `${modelToCreateDirectory}/${model.name}.js`
  const doesModelFileExist = fs.existsSync(modelToCreateFilePath)
  if (doesModelFileExist) {
    return
  }

  const constructorAssignments = model.propertyNames.map(p => {
    return `\nthis.${p} = props.${p}`
  })

  let constructorAssignmentsAsString = ''
  constructorAssignments.forEach(a => {
    constructorAssignmentsAsString += a
  })

  const getterAndSetterMethods = model.propertyNames.map(p => {
    return `
    get ${p} () {
        return this._${p}
      }
    
      set ${p} (value) {
        this._${p} = value
        return this._${p}
      }
    `
  })

  let getterAndSetterMethodsAsString = ''
  getterAndSetterMethods.forEach(m => {
    getterAndSetterMethodsAsString += m
  })

  const propsGetterObject = model.propertyNames.map(p => {
    return `\n${p}: this.${p},`
  })
  let propsGetterObjectAsString = ''
  propsGetterObject.forEach(p => {
    propsGetterObjectAsString += p
  })



  const modelClassText = `class ${model.name} {
      constructor (props = {}) {${constructorAssignmentsAsString}
      }
      ${getterAndSetterMethodsAsString}
      get props () {
        return {${propsGetterObjectAsString}
        }
      }
    }

    export default ${model.name}
  `

  fs.writeFileSync(modelToCreateFilePath, modelClassText, 'utf-8')
  vscode.window.showInformationMessage('Successfully Create Model')
}

export default createModel


src/UseCases/createModelTest.ts
import * as fs from 'fs'
import * as vscode from 'vscode'
import { v4 as uuidv4 } from 'uuid'

interface ModelInput {
  name: string,
  propertyNames: string[]
}

function createModelTest (model: ModelInput, context: vscode.ExtensionContext) {
  const workspaceDirectory: string = vscode.workspace.rootPath || ''

  const modelTestsDirectory = `${workspaceDirectory}/${context.globalState.get('modelTestsDirectory')}`
  try {
    const doesModelTestsDirectoryExist = fs.existsSync(modelTestsDirectory)
    if (!doesModelTestsDirectoryExist) fs.mkdirSync(modelTestsDirectory)
  } catch {
    vscode.window.showErrorMessage('Ensure path to Model Tests directory exists and is set in Settings')
    return
  }

  const modelTestToCreateDirectory = `${modelTestsDirectory}/${model.name}`
  const doesMoodelToCreateDirectoryExist = fs.existsSync(modelTestToCreateDirectory)
  if (!doesMoodelToCreateDirectoryExist) fs.mkdirSync(modelTestToCreateDirectory)

  const modelTestToCreateFilePath = `${modelTestToCreateDirectory}/${model.name}Test.js`
  const doesModelFileExist = fs.existsSync(modelTestToCreateFilePath)
  if (doesModelFileExist) {
    return
  }

  const inputGetAssignments = model.propertyNames.map(p => {
    return `\n${p}: '${uuidv4()}',`
  })

  let inputGetAssignmentsAsString = ''
  inputGetAssignments.forEach(a => {
    inputGetAssignmentsAsString += a
  })

  const inputSetAssignments = model.propertyNames.map(p => {
    return `\n${p}: '${uuidv4()}',`
  })

  let inputSetAssignmentsAsString = ''
  inputSetAssignments.forEach(a => {
    inputSetAssignmentsAsString += a
  })

  const modelClassText = `import ${model.name} from '../../../${context.globalState.get('modelsDirectory')}/${model.name}/${model.name}.js'

    const getPropsTest = () => {
      const input = {${inputGetAssignmentsAsString}\n}

      const expectedOutput = {${inputGetAssignmentsAsString}\n}

      const modelInstance = new ${model.name}(input)
      const modelProps = modelInstance.props

      if (JSON.stringify(modelProps) === JSON.stringify(expectedOutput)) return true
      else return false
    }
    

    const setPropsTest = () => {
      const input = {${inputGetAssignmentsAsString}\n}

      const modifiedData = {${inputSetAssignmentsAsString}\n}

      const modelInstance = new ${model.name}(input)
      for (let key in modifiedData) {
        modelInstance[key] = modifiedData[key]
      }
      const modelProps = modelInstance.props

      if (JSON.stringify(modelProps) === JSON.stringify(modifiedData)) return true
      else return false
    }

    export default [
      { name: 'Model | Get ${model.name} Properties', test: getPropsTest },
      { name: 'Model | Set ${model.name} Properties', test: setPropsTest },
    ]
  `

  fs.writeFileSync(modelTestToCreateFilePath, modelClassText, 'utf-8')
  vscode.window.showInformationMessage('Successfully Create Model Test')
}

export default createModelTest


src/Views/MVCForm.ts
import * as vscode from 'vscode'
import createModel from '../UseCases/createModel'

class MVCForm {
  public static currentPanel: MVCForm | undefined
  public static readonly viewType = 'mvcForm'
  private readonly _panel: vscode.WebviewPanel
  private readonly _extentionUri: vscode.Uri
  private _disposables: vscode.Disposable[] = []
  
  public static show (extentionUri: vscode.Uri) {
    const column = vscode.window.activeTextEditor
      ? vscode.window.activeTextEditor.viewColumn
      : undefined
    
      if (MVCForm.currentPanel) {
        MVCForm.currentPanel._panel.reveal(column)
      }

      const panel = vscode.window.createWebviewPanel(
        MVCForm.viewType,
        'MVC Form',
        column || vscode.ViewColumn.One,
        {
          enableScripts: true,
          localResourceRoots: [vscode.Uri.joinPath(extentionUri, 'viewResources')]
        }
      )

      MVCForm.currentPanel = new MVCForm(panel, extentionUri)
  }

  public static revive (panel: vscode.WebviewPanel, extentionUri: vscode.Uri) {
    MVCForm.currentPanel = new MVCForm(panel, extentionUri)
  }

  private constructor (panel: vscode.WebviewPanel, extentionUri: vscode.Uri) {
    this._panel = panel
    this._extentionUri = extentionUri

    this._update()

    this._panel.onDidDispose(() => this.dispose(), null, this._disposables)

    this._panel.onDidChangeViewState((e: any) => {
      if (this._panel.visible) this._update()
    },
    null,
    this._disposables
    )
  }

  public dispose () {
    MVCForm.currentPanel = undefined
    this._panel.dispose()
    
    while (this._disposables.length) {
      const x = this._disposables.pop()
      if (x) x.dispose()
    }
  }

  private _update () {
    const webview = this._panel.webview
    
    this._panel.title = 'MVC Generator'
    this._panel.webview.html = this._getHtmlForWebview(webview)
    this._panel.webview.onDidReceiveMessage((e: any) => {
      createModel(e)
    },
    undefined,
    this._disposables
    )
  }

  private _getHtmlForWebview(webview: vscode.Webview) {
    // const controllerOnDisk = vscode.Uri.joinPath(this._extentionUri, 'Controllers', 'mvcFormController.js')
    // const controllerUri = webview.asWebviewUri(controllerOnDisk)

    return `
    <!DOCTYPE html>
    <html lang="en">
    
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>MVC Generator</title>
    
      <style>
        button {
          color: white;
          background-color: rgb(75, 147, 255);
          border: solid rgb(49, 139, 255) 1px;
          border-radius: 4px;
          padding: 6px 10px;
          display: block;
          margin: 10px 0px;
        }
    
        label {
          display: block;
          margin-top: 6px;
        }
    
        input {
          display: block;
          background-color: rgba(255, 255, 255, 0.1);
          color: rgb(250, 250, 250);
          font-size: 16px;
          padding: 6px 10px;
          border: solid 1px rgb(196, 196, 196);
          margin-bottom: 6px;
        }
      </style>
    </head>
    
    <body>
      <label for="modelName"> Model Name</label>
      <input id="modelName" />
    
      <div id="propertyList">
        <h4>Properties</h4>
        <input class="propertyNameInput" />
      </div>
      <button onclick="addProperty()">Add Property</button>
    
      <br />
      <button onclick="submit()">Generate</button>
    </body>
    
    <script>
      const vscode = acquireVsCodeApi()
    
      function addProperty () {
        const input = document.createElement('input')
        input.setAttribute('class', 'propertyNameInput')
        
        const newPropertyElement = document.createElement('div')
        newPropertyElement.appendChild(input)
    
        const listElement = document.getElementById('propertyList')
        listElement.appendChild(newPropertyElement)
    
        input.focus()
      }
      
      function submit() {
        const inputs = document.querySelectorAll('.propertyNameInput') || []
        const propertyNames = [...inputs].map(i => i.value)
        
        vscode.postMessage(
          {
            name: document.getElementById('modelName').value,
            propertyNames: propertyNames
          }
        )
      }
    </script>
    
    </html>
    `
  }
}

export default MVCForm

